src/GraphQL/Types/ResolveFieldContextAdapter.cs

```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using GraphQL.Instrumentation;
+using GraphQL.Language.AST;
+
+namespace GraphQL.Types
+{
+    internal class ResolveFieldContextAdapter<T> : IResolveFieldContext<T>
+    {
+        private readonly IResolveFieldContext _baseContext;
+
+        public ResolveFieldContextAdapter(IResolveFieldContext baseContext)
+        {
+            _baseContext = baseContext ?? throw new ArgumentNullException(nameof(baseContext));
+
+            if (baseContext.Source is null)
+                Source = default(T);
```

Comment:
For value types this code will be misleading, converting null to default(valuetype). See my another comment about tests.

Reply:
True, but I'm not sure what you're suggesting.  Shall I add a comment?  Writing `if (typeof(T).IsValueType)` would have a significant performance impact when using typed IFieldResolvers.

Reply:
I am talking about runtime check not about comment.

> significant performance impact

Why do you think so? You can check this with benchmarks. I generally think that the JIT should expand this construction into a constant and delete the entire block of code or execute it unconditionally.